src/photon/photon_algorithm.c

```diff
@@ -355,22 +412,64 @@ void handle_object_available(local_scheduler_state *state,
     CHECK(event_loop_remove_timer(state->loop, fetch_req->timer) == AE_OK);
     free(fetch_req);
   }
+
+  /* Move any tasks whose object dependencies are now ready to the dispatch
+   * queue. */
+  /* TODO(swang): This can be optimized by keeping a lookup table from object
+   * ID to list of dependent tasks in the waiting queue. */
+  task_queue_entry *elt, *tmp;
+  DL_FOREACH_SAFE(algorithm_state->waiting_task_queue, elt, tmp) {
+    if (can_run(algorithm_state, elt->spec)) {
+      LOG_DEBUG("Moved task to dispatch queue");
+      DL_DELETE(algorithm_state->waiting_task_queue, elt);
+      DL_APPEND(algorithm_state->dispatch_task_queue, elt);
```

Comment:
Interesting, I assume you manually use `DL_APPEND` instead of calling `queue_task` to avoid issuing another TASK_TABLE_UPDATE?